### ISSUE TITLE
UI Design Epic and PipelineRun Interactive Diagram Implementation

### ISSUE THREAD

**Issue by kimholmes** [tektoncd/dashboard#675]
# Expected Behavior
- A user of the Tekton Dashboard needs to find what's failed in his PipelineRun at a glance. 
- A user of the Tekton Dashboard needs to drill into a failed step and quickly access the log file that corresponds to the failed attempted.
![PLRdiagram1sm](https://user-images.githubusercontent.com/21316329/67706212-c0649400-f98e-11e9-88cc-e304fb4caa5c.png)

# Actual Behavior
![Tek_Diagram_existing](https://user-images.githubusercontent.com/21316329/67693629-622cb680-f978-11e9-9e0d-a28beeeaf847.png)

# Additional Info
See the comments for high fidelity designs for the Tekton dashboard PipelienRun Diagram, based on the Carbon design system. It includes a map of the intended behaviors and HTML/CSS specs.

We can use the map of the behaviors to iterate on as FED works through the code.

**Comment by kimholmes** [comment]
View the high fidelity mockups at the following git links. Click on object to view associated specs:
### **PipelineRun Diagram View "Closed"**
https://kimholmes.github.io/Exported-Specs/PipelineTopology_Closed/
![PLRdiagram1](https://user-images.githubusercontent.com/21316329/67696702-b5553800-f97d-11e9-866b-97655e5f2ba9.png)

### **PipelineRun Diagram View "Open" - Node Selected**
https://kimholmes.github.io/Exported-Specs/PipelineTopology_DetailsLogs/
![PLRdiagram2](https://user-images.githubusercontent.com/21316329/67697429-e5e9a180-f97e-11e9-9052-8592f5ed8653.png)

### **PipelineRun Diagram View Logs "View more"**
https://kimholmes.github.io/Exported-Specs/PipelineTopology_LogsModal/
![PLRdiagram3](https://user-images.githubusercontent.com/21316329/67697735-66100700-f97f-11e9-9b5f-eee00de3942e.png)

**Comment by kimholmes** [comment]
### **Hover states:**
![PLR_DiagramHover](https://user-images.githubusercontent.com/21316329/67704251-218a6880-f98b-11e9-8fb2-5872dcf2f9f6.png)

### **Node selection:**
![PLR_DiagramNodeSelection](https://user-images.githubusercontent.com/21316329/67704505-9e1d4700-f98b-11e9-91fe-09dc95282773.png)

### **Accordion selection:**
![PLR_DiagramAccordionSelection](https://user-images.githubusercontent.com/21316329/67704808-31567c80-f98c-11e9-8155-6831740cef77.png)

### **Toolbar and Cursor behavior:**
![PLR_DiagramToolbarCursor](https://user-images.githubusercontent.com/21316329/67705163-dd986300-f98c-11e9-9e08-82d77f3212e7.png)

### **View Logs:**
![PLR_DiagramViewLogs](https://user-images.githubusercontent.com/21316329/67705495-7cbd5a80-f98d-11e9-8382-7e9ffa9e0cf5.png)

**Comment by kimholmes** [comment]
@AlanGreene

**Comment by AlanGreene** [comment]
@kimholmes How should we handle the hover state for expanded Tasks?

RIght now it just applies the same background fill to the whole node as when hovering in the collapsed state but it doesn't look right to me:
![image](https://user-images.githubusercontent.com/2829095/68896114-b5498c00-0722-11ea-9e53-b940d46af72e.png)

Do we want to apply this only to the line that's being hovered? Something like this maybe?
![image](https://user-images.githubusercontent.com/2829095/68896662-de1e5100-0723-11ea-81a5-52289b748d25.png)

**Comment by kimholmes** [comment]
@AlanGreene If you're able to apply it only to the row being hovered on, that would be ideal.

**Comment by AlanGreene** [comment]
Latest here: https://alangreene.github.io/dashboard/?path=/story/graph-graph--default
There are still some quirks to work out:
- excess whitespace below tasks with multiple steps
- some steps don't stretch the full width

We should have the ability to expand/collapse the tasks early tomorrow. I had hoped to have that complete today but ran into some issues.

**Comment by carrenelloyd** [comment]
### Changes related to the current Storybook implementation 
https://alangreene.github.io/dashboard/?path=/story/graph-pipelinegraph--default

### Height
**30 px - Height of 1 pipeline**
The icons + text inside the 30 pixel box should be centered vertically, meaning there should be the same amount of height above and below the text and icons. 
![image](https://user-images.githubusercontent.com/53481064/70167446-b8d29200-1694-11ea-8c3c-54ebb19ee7bc.png)
 
**60 px - Height of 2 pipelines**
![image](https://user-images.githubusercontent.com/53481064/70167505-def83200-1694-11ea-9a09-4ef285ba45e5.png)
 
**90 px - Height of 3 pipelines** 
The icons + text evenly distributed within the box
![image](https://user-images.githubusercontent.com/53481064/70169740-6647a480-1699-11ea-86ac-9f9ae9382299.png)
The pattern continues for each line of text: 120 px - Height of 4 pipelines, etc

Remove extra spacing at the bottom of the gray "Hasn't run yet" box
![image](https://user-images.githubusercontent.com/53481064/70169539-0cdf7580-1699-11ea-999b-d9eb080854e7.png)


### Arrows + Lines
The current implementation of the arrows looks like it has two pixels? If this is not the intention then perhaps it is not rendering pixels along the gird?
![image](https://user-images.githubusercontent.com/53481064/70167906-b6246c80-1695-11ea-8875-d71e2aa2652f.png)


The same goes for the arrows—it should end with only one point, not a double point
![image](https://user-images.githubusercontent.com/53481064/70168049-0d2a4180-1696-11ea-9da9-c3da5c028716.png)

- 1 px arrows
- $ui-04 color token
![image](https://user-images.githubusercontent.com/53481064/70167729-5928b680-1695-11ea-91b3-1c9931118e7a.png)

### Width
Indicated below with purple boxes, the spacing btwn the elements on the sections should be Token: $spacing-03
- This means there is not a fixed-width box size and the expand/collapse icon will not align to the right of the box like it does in the current implementation. The box will widen with an increase in characters. 
- **Set max-width to:** 200 px
- If the text exceeds the max-width of 200 pixels use an ellipsis in the center of the text with a [Carbon Tool Tip](https://www.carbondesignsystem.com/components/tooltip/code) with the entire string 
![image](https://user-images.githubusercontent.com/53481064/70168232-7316c900-1696-11ea-8a57-c9c2f22bceab.png)



### Color

The following elements should be using color Token $ui-04
1. Outline around box
2. Icon
3. Text
4. Expand/Collapse arrow
![image](https://user-images.githubusercontent.com/53481064/70169232-6e531480-1698-11ea-9fca-4d4e6349339b.png) 

### Icon size
The icon size that should be used is 20x20 this will help legibility with in the icon itself
![image](https://user-images.githubusercontent.com/53481064/70169401-bc681800-1698-11ea-9e63-795a8d5a2875.png)

**Comment by tomhau01** [comment]
Do you have any estimate when this work could be available for trying out? That flow chart type of visualization of steps would be really handy.

**Comment by AlanGreene** [comment]
@tomhau01 this is under active development, you can see the current status in the storybook https://alangreene.github.io/dashboard/?path=/story/graph-pipelinegraph--default. The PipelineGraph example uses Tekton resources (in JSON format) as input to generate the graph, so all of the basics are in place to build the data model and render the diagram.

Some updates are still needed to fully support inline resource definitions and some other features added in the most recent Tekton Pipeline release. We're also working on adding the ability to pan and zoom the diagram which will be important for larger / more complex pipelines.

Keep an eye on this issue for updates as we begin making changes to the PipelineRun details page and integrating the graph. The next couple of weeks should see most of the major changes complete.

**Comment by tomhau01** [comment]
Thank you for the update, and I'll be happy to give it a spin once there is something to try out 👍

**Comment by AlanGreene** [comment]
Next steps:
- review compatibility with current pipelines releases
  - any impact on how links in the DAG are constructed based on new features?
  - probably OK to ignore Conditions since they're deprecated and being replaced by 'when expressions'
  - how should we display retries? Worth considering this along with https://github.com/tektoncd/dashboard/issues/1445, may have some impact
- resolve styling issues
  - unwanted whitespace in tasks with multiple steps when expanded
- decide on how the graph will be presented
  - potentially as an overlay / side panel initially until other design update on the PipelineRun + TaskRun pages can be implemented

**Comment by AlanGreene** [comment]
> - resolve styling issues
>   - unwanted whitespace in tasks with multiple steps when expanded

If this can't be resolved using the current approach, it may be worth considering alternative layout algorithms, or possibly changing the Node design (removing the expand/collapse behaviour) if this is contributing to the problem.

One pattern that seems to be pretty common is having a simplified summary of the child nodes displayed permanently, with more detailed information displayed in a tooltip / popover of some sort on hover or other interaction.

**Comment by AlanGreene** [comment]
Potential updated design:
![image](https://user-images.githubusercontent.com/2829095/105485336-a0f90b80-5ca4-11eb-83a1-e3c9513ce6fe.png)

It's more compact and slightly less likely to have unwanted edge crossings.

This requires only a small config change to the current graph code and some minor style tweaks.

Current POC keeps the expanding Task behaviour from the original:
![image](https://user-images.githubusercontent.com/2829095/105485507-e74e6a80-5ca4-11eb-858a-57ee466aa487.png)

There are still layout issues with this in some cases, but they may be easily solved. If not we can adopt a different approach to surface the step information, either directly on the Task nodes or in some form of tooltip.

**Comment by christianvogt** [comment]
Is this visualization available in the dashboard or when is it planned?

The [OpenShift Console](https://github.com/openshift/console) and the [Tekton Pipelines VSCode Extension](https://github.com/redhat-developer/vscode-tekton) have visualizations. It would be great if a single pipeline visualization library could be shared amongst all 3 tools by creating a separate reusable library.

**Comment by AlanGreene** [comment]
I wasn't aware of the preview diagram provided by the VSCode extension, I'll have to take a look at that. Thanks for sharing.

The visualisation is not currently available in the Dashboard but we have revisited this issue recently (~1 month) and the latest design explorations using [Carbon Charts](https://github.com/carbon-design-system/carbon-charts) to fit with the Carbon design language used throughout the rest of the Dashboard can be found here: https://54h4e.csb.app/

We're currently in the middle of a significant redesign effort for the PipelineRun list + details views and hope to have more time to dedicate to the visualisation work once other high priority items are complete. It's still something we're eager to add to the Dashboard.

**Comment by number5** [comment]
@AlanGreene just FYI,  https://54h4e.csb.app/ is broken
```
Error
Undefined variable.
  ╷
1 │ .#{$prefix}--#{$charts-prefix}--grid {
  │    ^^^^^^^
  ╵
  stdin 1:4  root stylesheet
```

**Comment by AlanGreene** [comment]
@number5 thanks for the heads up, looks like an intermittent issue with the Codesandbox environment itself. We haven't changed the code in over a month and it's working for me at the moment. I did hit that error above a couple of times and it looked like Codesandbox hadn't fully installed the dependencies or was having trouble reading from the filesystem.

I'll try to narrow it down and open an issue with the Codesandbox folks.

**Comment by pawelkopka** [comment]
Hi, is there any eta?

**Comment by AlanGreene** [comment]
Apologies for the lack of updates. There's no ETA right now as we have been concentrating on other priorities. I have started work on migrating the POC described above into the Dashboard repo so we can begin reviewing and iterating on it properly, I'll post updates here or in linked PRs as they happen. I expect to have the initial migration complete within the next week or two. After that we will begin to review the changes required to support features introduced in recent Tekton Pipelines releases (e.g. matrix, when expressions, custom tasks, etc.) and create sub-issues to track them as appropriate.

If you have specific use cases in mind and how a visualisation could help you to achieve them, please share details as it would be very helpful to us as we finalise the designs.

**Comment by seanlaff** [comment]
@AlanGreene Thanks for the update. We're coming from a gitlab environment where this dag-pipeline views are invaluable for a large complex repo. 

With regards to suggestions- I love if there was some way to surface the running-time of each task on the node. It would help us quickly diagnose bottlenecks in the pipeline.

**Comment by Goltergaul** [comment]
Is there any news on this?

**Comment by cmorinupgrade** [comment]
@AlanGreene Is this feature on the roadmap? That's one feature gap with Argo Workflows, which would drastically improve developer experience with regards to seeing how the DAG looks like and how it progresses through it. Would love to see such a feature added to the dashboard, among other developer experience focused improvements.

**Comment by AlanGreene** [comment]
Yes this is still something we'd like to add.

**Comment by cmorinupgrade** [comment]
Do we have any ETA in mind?

**Comment by valAndre07** [comment]
Hello @AlanGreene  any ETA about this very useful feature ?

**Comment by valAndre07** [comment]
Hello, is there any progress about this very useful feature?

**Comment by valAndre07** [comment]
Can I help on this one ?

**Comment by AlanGreene** [comment]
Thanks for the interest. As mentioned above, the most valuable contributions at the moment would be details of specific use cases for the visualisation.

What tasks are users unable to complete without it, or what tasks would be made easier / more efficient by its introduction?

What problems are users trying to solve or what questions are they trying to answer?

These are crucial to ensure we understand the requirements, are solving the right problems, and aren't just adding a visualisation for the sake of it.

Once we have a proper understanding of the requirements we can propose designs to address them, including how best to integrate the visualisation into existing pages / flows, any additional functionality it must support, etc.

**Comment by oskarnrk** [comment]
Hi! I think this is a great addition. It makes it easy to see a PipelineRun’s status at a glance and the visualized dependencies between tasks are really helpful for understanding failure impact. Being able to drill into a failed step and jump straight to the corresponding log should speed up debugging a lot. I used to see a similar diagram on OpenShift and it helped me a lot.

**Comment by mcodato** [comment]
<img width="2521" height="644" alt="Image" src="https://github.com/user-attachments/assets/7bc744e0-d273-428b-b9fc-4444a5cd1591" />

Hi all!

To add a concrete example of why this would be valuable: below is one of our more complex PipelineRuns. It has several build tasks running in parallel, which fan into a deployment step, followed by another parallel wave of tasks that converge again at the end.

Reading this as a flat list or YAML makes the dependency structure very hard to follow. The graph makes the fan-out/fan-in pattern immediately obvious, and you can see at a glance which tasks gate others and where the critical path runs.

Beyond what @oskarnrk mentioned about debugging failures, it also helps during pipeline development: it's a quick way to check that the dependency wiring gives you the intended parallelism, rather than accidentally serializing tasks that could run concurrently.

For context, this graph comes from our OpenShift Pipelines setup. We're currently moving away from OpenShift, so having this visualization available in the Tekton Dashboard would be a big help for us to keep the same level of insight.

Thanks all for the great work on this project  :)

**Comment by kimholmes** [comment]
OMG. I opened this issue 7 years ago! lol.
